File: app/src/main/java/org/ole/planet/myplanet/base/BaseRecyclerFragment.java
package org.ole.planet.myplanet.base;

import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.ole.planet.myplanet.R;
import org.ole.planet.myplanet.callback.OnRatingChangeListener;
import org.ole.planet.myplanet.datamanager.DatabaseService;
import org.ole.planet.myplanet.model.RealmCourseProgress;
import org.ole.planet.myplanet.model.RealmMyCourse;
import org.ole.planet.myplanet.model.RealmMyLibrary;
import org.ole.planet.myplanet.model.RealmRemovedLog;
import org.ole.planet.myplanet.model.RealmStepExam;
import org.ole.planet.myplanet.model.RealmSubmission;
import org.ole.planet.myplanet.model.RealmTag;
import org.ole.planet.myplanet.service.UserProfileDbHandler;
import org.ole.planet.myplanet.utilities.Utilities;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import io.realm.Case;
import io.realm.RealmList;
import io.realm.RealmObject;

import static android.content.Context.MODE_PRIVATE;

public abstract class BaseRecyclerFragment<LI> extends BaseRecyclerParentFragment implements OnRatingChangeListener {
    public static final String PREFS_NAME = "OLE_PLANET";
    public static SharedPreferences settings;
    public Set<String> subjects, languages, mediums, levels;
    public List<LI> selectedItems;
    public String gradeLevel = "", subjectLevel = "";
    public DatabaseService realmService;
    public UserProfileDbHandler profileDbHandler;

    public RecyclerView recyclerView;
    public TextView tvMessage, tvFragmentInfo;
    public TextView tvDelete;
    List<LI> list;

    public BaseRecyclerFragment() {
    }

    public static void showNoData(View v, int count) {
        if (v == null)
            return;
        v.setVisibility(count == 0 ? View.VISIBLE : View.GONE);
        ((TextView) v).setText("No data available, please check and try again.");
    }

    public abstract int getLayout();

    public abstract RecyclerView.Adapter getAdapter();

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            isMyCourseLib = getArguments().getBoolean("isMyCourseLib");
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(getLayout(), container, false);
        settings = getActivity().getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
        recyclerView = v.findViewById(R.id.recycler);
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        if (isMyCourseLib) {
            tvDelete = v.findViewById(R.id.tv_delete);
            initDeleteButton();
            if (v.findViewById(R.id.tv_add) != null)
                v.findViewById(R.id.tv_add).setVisibility(View.GONE);

        }
        tvMessage = v.findViewById(R.id.tv_message);
        selectedItems = new ArrayList<>();
        list = new ArrayList<>();
        realmService = new DatabaseService(getActivity());
        mRealm = realmService.getRealmInstance();
        profileDbHandler = new UserProfileDbHandler(getActivity());
        model = profileDbHandler.getUserModel();
        recyclerView.setAdapter(getAdapter());
        if (isMyCourseLib) showDownloadDialog(getLibraryList(mRealm));
        return v;
    }

    protected void initDeleteButton() {
        if (tvDelete != null) {
            tvDelete.setVisibility(View.VISIBLE);
            tvDelete.setOnClickListener(view -> deleteSelected(false));
        }
    }

    @Override
    public void onRatingChanged() {
        recyclerView.setAdapter(getAdapter());
    }

    public void addToMyList() {
        for (int i = 0; i < selectedItems.size(); i++) {
            RealmObject object = (RealmObject) selectedItems.get(i);
            if (object instanceof RealmMyLibrary) {
                RealmMyLibrary myObject = mRealm.where(RealmMyLibrary.class).equalTo("resourceId", ((RealmMyLibrary) object).getResource_id()).findFirst();
                RealmMyLibrary.createFromResource(myObject, mRealm, model.getId());
                RealmRemovedLog.onAdd(mRealm, "resources", profileDbHandler.getUserModel().getId(), myObject.getResourceId());
                Utilities.toast(getActivity(), "Added to my library");
                recyclerView.setAdapter(getAdapter());
            } else {
                RealmMyCourse myObject = RealmMyCourse.getMyCourse(mRealm, ((RealmMyCourse) object).getCourseId());
                RealmMyCourse.createMyCourse(myObject, mRealm, model.getId());
                RealmRemovedLog.onAdd(mRealm, "courses", profileDbHandler.getUserModel().getId(), myObject.getCourseId());
                Utilities.toast(getActivity(), "Added to my courses");
                recyclerView.setAdapter(getAdapter());
            }
            showNoData(tvMessage, getAdapter().getItemCount());
        }
    }

    public void deleteSelected(boolean deleteProgress) {
        for (int i = 0; i < selectedItems.size(); i++) {
            if (!mRealm.isInTransaction())
                mRealm.beginTransaction();
            RealmObject object = (RealmObject) selectedItems.get(i);
            deleteCourseProgress(deleteProgress, object);
            removeFromShelf(object);
            recyclerView.setAdapter(getAdapter());
            showNoData(tvMessage, getAdapter().getItemCount());
        }
    }

    private void deleteCourseProgress(boolean deleteProgress, RealmObject object) {
        if (deleteProgress && object instanceof RealmMyCourse) {
            mRealm.where(RealmCourseProgress.class).equalTo("courseId", ((RealmMyCourse) object).getCourseId()).findAll().deleteAllFromRealm();
            List<RealmStepExam> examList = mRealm.where(RealmStepExam.class).equalTo("courseId", ((RealmMyCourse) object).getCourseId()).findAll();
            for (RealmStepExam exam : examList) {
                mRealm.where(RealmSubmission.class)
                        .equalTo("parentId", exam.getId())
                        .notEqualTo("type", "survey")
                        .equalTo("uploaded", false).findAll().deleteAllFromRealm();
            }
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mRealm.close();
    }

    private void checkAndAddToList(RealmMyCourse course, List<RealmMyCourse> courses, List<RealmTag> tags) {
        for (RealmTag tg : tags) {
            long count = mRealm.where(RealmTag.class).equalTo("db", "courses").equalTo("tagId", tg.getId()).equalTo("linkId", course.getCourseId()).count();
            if (count > 0 && !courses.contains(course))
                courses.add(course);
        }
    }

    private List<LI> getData(String s, Class c) {
        List<LI> li = new ArrayList<>();
        if (!s.contains(" ")) {
            li = mRealm.where(c).contains(c == RealmMyLibrary.class ? "title" : "courseTitle", s, Case.INSENSITIVE).findAll();
        } else {
            String[] query = s.split(" ");
            List<LI> data = mRealm.where(c).findAll();
            for (LI l : data) {
                searchAndAddToList(l, c, query, li);
            }
        }
        return li;
    }

    private void searchAndAddToList(LI l, Class c, String[] query, List<LI> li) {
        String title = c == RealmMyLibrary.class ? ((RealmMyLibrary) l).getTitle() : ((RealmMyCourse) l).getCourseTitle();
        boolean isExists = false;
        for (String q : query) {
//                li.add(l);
            isExists = title.toLowerCase().contains(q.toLowerCase());
            Utilities.log(title.toLowerCase() + " " + q.toLowerCase() + " is exists " + isExists);

            if (!isExists) break;
        }
        if (isExists)
            li.add(l);
    }

    public List<RealmMyLibrary> filterLibraryByTag(String s, List<RealmTag> tags) {
        if (tags.size() == 0 && s.isEmpty()) {
            return (List<RealmMyLibrary>) getList(RealmMyLibrary.class);
        }
        List<RealmMyLibrary> list = (List<RealmMyLibrary>) getData(s, RealmMyLibrary.class);
        if (isMyCourseLib) list = RealmMyLibrary.getMyLibraryByUserId(model.getId(), list);
        else list = RealmMyLibrary.getOurLibrary(model.getId(), list);
        if (tags.size() == 0) return list;
        RealmList<RealmMyLibrary> libraries = new RealmList<>();
        for (RealmMyLibrary library : list) {
            filter(tags, library, libraries);
        }
        return libraries;
    }


    public List<RealmMyCourse> filterCourseByTag(String s, List<RealmTag> tags) {
        if (tags.size() == 0 && s.isEmpty()) {
            return applyCourseFilter((List<RealmMyCourse>) getList(RealmMyCourse.class));
        }
        List<RealmMyCourse> list = (List<RealmMyCourse>) getData(s, RealmMyCourse.class);
        if (isMyCourseLib) list = RealmMyCourse.getMyCourseByUserId(model.getId(), list);
        else list = RealmMyCourse.getOurCourse(model.getId(), list);
        if (tags.size() == 0) return list;
        RealmList<RealmMyCourse> courses = new RealmList<>();
        for (RealmMyCourse course : list) {
            checkAndAddToList(course, courses, tags);
        }
        return applyCourseFilter(list);
    }

    private void filter(List<RealmTag> tags, RealmMyLibrary library, RealmList<RealmMyLibrary> libraries) {
        for (RealmTag tg : tags) {
            long count = mRealm.where(RealmTag.class).equalTo("db", "resources").equalTo("tagId", tg.getId()).equalTo("linkId", library.getId()).count();
            if (count > 0 && !libraries.contains(library))
                libraries.add(library);
        }
    }

    public List<RealmMyLibrary> applyFilter(List<RealmMyLibrary> libraries) {
        List<RealmMyLibrary> newList = new ArrayList<>();
        for (RealmMyLibrary l : libraries) {
            if (isValidFilter(l)) newList.add(l);
        }
        return newList;
    }

    public List<RealmMyCourse> applyCourseFilter(List<RealmMyCourse> courses) {
        Utilities.log("apply course filter");
        if (TextUtils.isEmpty(subjectLevel) && TextUtils.isEmpty(gradeLevel))
            return courses;
        List<RealmMyCourse> newList = new ArrayList<>();
        for (RealmMyCourse l : courses) {
            Utilities.log("grade " + gradeLevel);
            Utilities.log("subject " + subjectLevel);
            if (TextUtils.equals(l.getGradeLevel(), gradeLevel) || TextUtils.equals(l.getSubjectLevel(), subjectLevel)) {
                newList.add(l);
            }
        }
        return newList;
    }

    private boolean isValidFilter(RealmMyLibrary l) {
        boolean sub = subjects.isEmpty() || l.getSubject().containsAll(subjects);
        boolean lev = levels.isEmpty() || l.getLevel().containsAll(levels);
        boolean lan = languages.isEmpty() || languages.contains(l.getLanguage());
        boolean med = mediums.isEmpty() || mediums.contains(l.getMediaType());
        return (sub && lev && lan && med);
    }
}

File: app/src/main/java/org/ole/planet/myplanet/base/BaseRecyclerFragment.kt
package org.ole.planet.myplanet.base

import android.os.Build
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import io.realm.RealmList
import io.realm.RealmObject
import java.text.Normalizer
import java.util.Locale
import kotlinx.coroutines.launch
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.callback.OnRatingChangeListener
import org.ole.planet.myplanet.model.RealmMyCourse
import org.ole.planet.myplanet.model.RealmMyLibrary
import org.ole.planet.myplanet.model.RealmTag
import org.ole.planet.myplanet.utils.Utilities.toast

abstract class BaseRecyclerFragment<LI> : BaseRecyclerParentFragment<Any?>(), OnRatingChangeListener {
    var subjects: MutableSet<String> = mutableSetOf()
    var languages: MutableSet<String> = mutableSetOf()
    var mediums: MutableSet<String> = mutableSetOf()
    var levels: MutableSet<String> = mutableSetOf()
    var selectedItems: MutableList<LI>? = null
    var gradeLevel = ""
    var subjectLevel = ""
    lateinit var recyclerView: RecyclerView
    var tvMessage: View? = null
    lateinit var tvFragmentInfo: TextView
    var tvDelete: TextView? = null
    var list: MutableList<LI>? = null
    var resources: List<RealmMyLibrary>? = null
    var courseLib: String? = null
    private var isAddInProgress = false


    abstract fun getLayout(): Int

    abstract suspend fun getAdapter(): RecyclerView.Adapter<out RecyclerView.ViewHolder>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            isMyCourseLib = it.getBoolean("isMyCourseLib")
            courseLib = it.getString("courseLib")
            @Suppress("UNCHECKED_CAST")
            resources = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                it.getSerializable("resources", ArrayList::class.java) as? ArrayList<RealmMyLibrary>
            } else {
                @Suppress("DEPRECATION")
                it.getSerializable("resources") as? ArrayList<RealmMyLibrary>
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(getLayout(), container, false)
        recyclerView = v.findViewById(R.id.recycler)
        recyclerView.layoutManager = LinearLayoutManager(activity)
        if (isMyCourseLib) {
            tvDelete = v.findViewById(R.id.tv_delete)
            initDeleteButton()
            v.findViewById<TextView>(R.id.tv_add)?.visibility = View.GONE
        }
        tvMessage = v.findViewById(R.id.tv_message)
        selectedItems = mutableListOf()
        list = mutableListOf()
        return v
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        postponeEnterTransition()
        viewLifecycleOwner.lifecycleScope.launch {
            requireRealmInstance()
            model = profileDbHandler.getUserModel()
            val adapter = getAdapter()
            recyclerView.adapter = adapter
            if (isMyCourseLib && adapter.itemCount != 0 && courseLib == "courses") {
                resources?.let { showDownloadDialog(it) }
            } else if (isMyCourseLib && courseLib == null && !isSurvey) {
                viewLifecycleOwner.lifecycleScope.launch {
                    val userId = sharedPrefManager.getUserId().ifEmpty { "--" }
                    val libraryList = resourcesRepository.getLibraryListForUser(userId)
                    showDownloadDialog(libraryList)
                }
            }
            startPostponedEnterTransition()
            requireActivity().reportFullyDrawn()
        }
    }

    private fun initDeleteButton() {
        tvDelete?.let {
            it.visibility = View.VISIBLE
            it.setOnClickListener {
                viewLifecycleOwner.lifecycleScope.launch {
                    deleteSelected(false)
                }
            }
        }
    }

    override fun onRatingChanged() {
        viewLifecycleOwner.lifecycleScope.launch {
            recyclerView.adapter = getAdapter()
        }
    }

    open fun addToMyList() {
        if (!isRealmInitialized() || isAddInProgress) return

        val itemsToAdd = selectedItems?.toList() ?: emptyList()
        if (itemsToAdd.isEmpty()) return

        val resourceIds = mutableListOf<String>()
        val courseIds = mutableListOf<String>()

        itemsToAdd.forEach { item ->
            when (val realmObject = item as? RealmObject) {
                is RealmMyLibrary -> realmObject.resourceId?.let(resourceIds::add)
                is RealmMyCourse -> realmObject.courseId?.let(courseIds::add)
                else -> {}
            }
        }

        if (resourceIds.isEmpty() && courseIds.isEmpty()) return

        isAddInProgress = true
        setJoinInProgress(true)

        viewLifecycleOwner.lifecycleScope.launch {
            val userId = profileDbHandler.getUserModel()?.id ?: return@launch
            var libraryAdded = false
            var courseAdded = false
            var errorOccurred: Throwable? = null

            if (resourceIds.isNotEmpty()) {
                val libraryResult = resourcesRepository.addResourcesToUserLibrary(resourceIds, userId)
                libraryResult.onSuccess {
                    libraryAdded = true
                }.onFailure {
                    errorOccurred = it
                }
            }

            if (courseIds.isNotEmpty()) {
                courseIds.forEach { courseId ->
                    val courseResult = coursesRepository.markCourseAdded(courseId, userId)
                    courseResult.onSuccess { added ->
                        if (added) {
                            courseAdded = true
                        }
                    }.onFailure {
                        errorOccurred = it
                    }
                }
            }

            isAddInProgress = false
            setJoinInProgress(false)

            if (view == null || !isAdded || requireActivity().isFinishing) return@launch

            if (!requireRealmInstance().isClosed) {
                requireRealmInstance().refresh()
            }

            val newAdapter = getAdapter()
            recyclerView.adapter = newAdapter
            showNoData(tvMessage, newAdapter.itemCount, "")

            errorOccurred?.let {
                it.printStackTrace()
                toast(activity, "An error occurred: ${it.message}")
                return@launch
            }

            if (libraryAdded) toast(activity, getString(R.string.added_to_my_library))
            if (courseAdded) toast(activity, getString(R.string.added_to_my_courses))
        }
    }

    private fun setJoinInProgress(inProgress: Boolean) {
        recyclerView.isEnabled = !inProgress
        recyclerView.alpha = if (inProgress) 0.6f else 1f
        view?.findViewById<View>(R.id.tv_add)?.let { addButton ->
            addButton.isEnabled = if (inProgress) {
                false
            } else {
                !(selectedItems.isNullOrEmpty())
            }
            addButton.alpha = if (inProgress) 0.5f else 1f
        }
    }

    open suspend fun deleteSelected(deleteProgress: Boolean) {
        selectedItems?.forEachIndexed { _, item ->
            val `object` = item as RealmObject
            deleteCourseProgress(deleteProgress, `object`)
            removeFromShelf(`object`)
        }
        selectedItems?.clear()
    }

    fun countSelected(): Int {
        return selectedItems?.size ?: 0
    }

    private suspend fun deleteCourseProgress(deleteProgress: Boolean, `object`: RealmObject) {
        if (deleteProgress && `object` is RealmMyCourse) {
            coursesRepository.deleteCourseProgress(`object`.courseId)
        }
    }

    fun normalizeText(str: String): String {
        return Normalizer.normalize(str.lowercase(Locale.getDefault()), Normalizer.Form.NFD)
            .replace(DIACRITICS_REGEX, "")
    }

    suspend fun filterCourseByTag(s: String, tags: List<RealmTag>): List<RealmMyCourse> {
        if (tags.isEmpty() && s.isEmpty()) {
            return applyCourseFilter(filterRealmMyCourseList(getList(RealmMyCourse::class.java) as List<RealmMyCourse>))
        }
        var list = coursesRepository.search(s)
        list = if (isMyCourseLib) {
            coursesRepository.getMyCourses(model?.id, list)
        } else {
            coursesRepository.getAllCourses(model?.id, list)
        }
        if (tags.isEmpty()) {
            return list
        }

        val tagIds = tags.mapNotNull { it.id }.toTypedArray()
        val linkedCourseIds = requireRealmInstance().where(RealmTag::class.java)
            .equalTo("db", "courses")
            .`in`("tagId", tagIds)
            .findAll()
            .mapNotNull { it.linkId }
            .toSet()

        val courses = RealmList<RealmMyCourse>()
        list.forEach { course ->
            if (linkedCourseIds.contains(course.courseId) && !courses.contains(course)) {
                courses.add(course)
            }
        }
        return applyCourseFilter(courses)
    }

    private fun filterRealmMyCourseList(items: List<Any?>): List<RealmMyCourse> {
        return items.filterIsInstance<RealmMyCourse>()
    }

    fun applyFilter(libraries: List<RealmMyLibrary>): List<RealmMyLibrary> {
        val newList: MutableList<RealmMyLibrary> = ArrayList()
        for (l in libraries) {
            if (isValidFilter(l)) newList.add(l)
        }
        return newList
    }

    private fun applyCourseFilter(courses: List<RealmMyCourse>): List<RealmMyCourse> {
        if (TextUtils.isEmpty(subjectLevel) && TextUtils.isEmpty(gradeLevel)) return courses
        val newList: MutableList<RealmMyCourse> = ArrayList()
        for (l in courses) {
            if (TextUtils.equals(l.gradeLevel, gradeLevel) || TextUtils.equals(
                    l.subjectLevel, subjectLevel
                )
            ) {
                newList.add(l)
            }
        }
        return newList
    }

    private fun isValidFilter(l: RealmMyLibrary): Boolean {
        val sub = subjects.isEmpty() || subjects.let { l.subject?.containsAll(it) } == true
        val lev = levels.isEmpty() || l.level?.containsAll(levels) == true
        val lan = languages.isEmpty() || languages.contains(l.language)
        val med = mediums.isEmpty() || mediums.contains(l.mediaType)
        return sub && lev && lan && med
    }

    override fun onDetach() {
        super.onDetach()
        cleanupReferences()
    }

    private fun cleanupReferences() {
        selectedItems?.clear()
        list?.clear()
        selectedItems = null
        list = null
        resources = null
    }

    companion object {
        private val DIACRITICS_REGEX = Regex("\\p{InCombiningDiacriticalMarks}+")

        private val noDataMessages = mapOf(
            "courses" to R.string.no_courses,
            "resources" to R.string.no_resources,
            "resources_filtered" to R.string.no_results_for_filter,
            "finances" to R.string.no_finance_record,
            "news" to R.string.no_voices_available,
            "news_filtered" to R.string.no_results_for_filter,
            "teamCourses" to R.string.no_team_courses,
            "teamResources" to R.string.no_team_resources,
            "tasks" to R.string.no_tasks,
            "members" to R.string.no_join_request_available,
            "discussions" to R.string.no_news,
            "survey" to R.string.no_surveys,
            "survey_submission" to R.string.no_survey_submissions,
            "exam_submission" to R.string.no_exam_submissions,
            "team" to R.string.no_teams,
            "enterprise" to R.string.no_enterprise,
            "chatHistory" to R.string.no_chats,
            "feedback" to R.string.no_feedback,
            "reports" to R.string.no_reports
        )

        fun showNoData(v: View?, count: Int?, source: String) {
            v ?: return
            v.visibility = if (count == 0) View.VISIBLE else View.GONE
            val messageRes = noDataMessages[source]
                ?: R.string.no_data_available_please_check_and_try_again
            val textView = if (v is TextView) v else v.findViewById<TextView>(R.id.tv_empty_message)
            textView?.setText(messageRes)
        }

        fun showNoFilter(v: View?, count: Int) {
            v ?: return
            v.visibility = if (count == 0) View.VISIBLE else View.GONE
            val textView = if (v is TextView) v else v.findViewById<TextView>(R.id.tv_empty_message)
            textView?.setText(R.string.no_course_matched_filter)
        }
    }
}
